### ISSUE TITLE
False warnings with dynamic class in match statement

### ISSUE THREAD

**Issue by ljodal** [microsoft/pyright#3520]
**Describe the bug**
Using a dynamically created class in a match statement produces a warning about it not being a class. I've attached some example code below that reproduces the issue. At runtime the code executes as expected.

In [my project](https://github.com/ljodal/hap) I'm getting a slightly different error (not sure why):

```
"State" cannot be used in a class pattern because it is a specialized type alias
```

**To Reproduce**

Check the reproducer below with pyright.

**Expected behavior**
I expect no warnings from the example code as it is correctly typed and works fine at runtime.

**Code**

```python
# reproducer.py
class Example:
    __match_args__ = ("_value",)

    def __init__(self, value: str) -> None:
        self._value = value

def example(name: str) -> type[Example]:
    return type(name, (Example,), {})

SubClass = example("SubClass")

t = [SubClass("my value"), SubClass("Other value")]

match t:
    case [SubClass(value), *rest]:
        print(f"Subclass with value: {value}")
```

```
→ npx pyright reproducer.py
No configuration file found.
No pyproject.toml file found.
stubPath <snip>/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
<snip>/reproducer.py
  <snip>/reproducer.py:15:11 - error: "Type[Example]" is not a class (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 0.431sec
```

**VS Code extension or command-line**
I'm seeing this problem both on the command line (running version `1.1.249` through `npx`) and in Vim with `coc-pyright` (which seem to be running a slightly older version).

**Additional context**
Mypy has a similar problem, see python/mypy#12886.

**Comment by erictraut** [comment]
Thanks for the bug report. When I implemented the support for match, I was using an earlier (alpha) implementation of Python 3.10 to validate its behavior. At that time, the interpreter generated a runtime error in this case, so I added an explicit error check. It appears that the final version of 3.10 supports this case without generating a runtime error, so I'll remove the corresponding check in pyright. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.250, which I just published. It will also be included in the next release of pylance.
